src/hud.cpp

```diff
@@ -464,15 +503,71 @@ void Hud::drawCrosshair() {
 	}
 }
 
-
-void Hud::drawSelectionBoxes(std::vector<aabb3f> &hilightboxes) {
-	for (std::vector<aabb3f>::const_iterator
-			i = hilightboxes.begin();
-			i != hilightboxes.end(); ++i) {
-		driver->draw3DBox(*i, selectionbox_argb);
+void Hud::setHighlightedPos(const v3f &pos, const v3s16 &camera_offset)
+{
+	m_camera_offset = camera_offset;
+	m_highlighted_pos = pos;
+	m_highlighted_pos_with_offset = pos - intToFloat(camera_offset, BS);
+}
+	
+void Hud::drawSelectionMesh()
+{	
+	if (!m_use_selectionmesh) {
+		video::SMaterial oldmaterial = driver->getMaterial2D();
+		driver->setMaterial(m_selection_material);
+		for (std::vector<aabb3f>::const_iterator
+				i = m_selectionboxes.begin();
+				i != m_selectionboxes.end(); ++i) {
+			aabb3f box = aabb3f(
+				i->MinEdge + m_highlighted_pos_with_offset,
+				i->MaxEdge + m_highlighted_pos_with_offset);
+			driver->draw3DBox(box, selectionbox_argb);
+		}
+		driver->setMaterial(oldmaterial);
+	} else if (m_selectionmesh) {
+		video::SMaterial oldmaterial = driver->getMaterial2D();
+		driver->setMaterial(m_selection_material);
+		setMeshColor(m_selectionmesh, m_highlightedmeshcolor);
+		scene::IMesh* mesh = cloneMesh(m_selectionmesh);
+		translateMesh(mesh, m_highlighted_pos_with_offset);
+		u32 mc = m_selectionmesh->getMeshBufferCount();
+		for (u32 i = 0; i < mc; i++) {
+			scene::IMeshBuffer *buf = mesh->getMeshBuffer(i);
+			driver->drawMeshBuffer(buf);
+		}
+		mesh->drop();
+		driver->setMaterial(oldmaterial);
 	}
 }
 
+void Hud::updateSelectionMesh(const v3s16 &camera_offset)
+{
+	m_camera_offset = camera_offset;
+	if (!m_use_selectionmesh)
+		return;
+	if (!m_selectionboxes.size()) {
+		if (m_selectionmesh) {
+			m_selectionmesh->drop();
+			m_selectionmesh = NULL;
+		}
+		return;
+	} else {
+		if (m_selectionmesh) {
+			m_selectionmesh->drop();
+			m_selectionmesh = NULL;
+		}
+		static f32 txc[24] = {
+			0,0,1,1,
+			0,0,1,1,
+			0,0,1,1,
+			0,0,1,1,
+			0,0,1,1,
+			0,0,1,1
+		};
+		m_selectionmesh = convertNodeboxesToMesh(m_selectionboxes, txc);
+		scaleMesh(m_selectionmesh, m_selectionmeshscale * m_overallselectionmeshscale);
```

Comment:
Probably a good idea to add some more underscores in these variable names here.  They're way too mushed together.